pkg/runtime/pubsub/outbox.go

```diff
@@ -0,0 +1,305 @@
+	lock                  sync.RWMutex
+}
+
+// NewOutbox returns an instance of an Outbox.
+func NewOutbox(publishFn func(context.Context, *contribPubsub.PublishRequest) error, getPubsubFn func(string) (contribPubsub.PubSub, bool), getStateFn func(string) (state.Store, bool), cloudEventExtractorFn func(map[string]any, string) string) outbox.Outbox {
+	return &outboxImpl{
+		cloudEventExtractorFn: cloudEventExtractorFn,
+		getPubsubFn:           getPubsubFn,
+		getStateFn:            getStateFn,
+		publishFn:             publishFn,
+		lock:                  sync.RWMutex{},
+		outboxStores:          map[string]outboxConfig{},
+	}
+}
+
+// AddOrUpdateOutbox examines a statestore for outbox properties and saves it for later usage in outbox operations.
+func (o *outboxImpl) AddOrUpdateOutbox(stateStore v1alpha1.Component) {
+	var publishPubSub, publishTopicKey, outboxPubsub, stateScanDelay string
+
+	for _, v := range stateStore.Spec.Metadata {
+		switch v.Name {
+		case outboxPublishPubsubKey:
+			publishPubSub = v.Value.String()
+		case outboxPublishTopicKey:
+			publishTopicKey = v.Value.String()
+		case outboxPubsubKey:
+			outboxPubsub = v.Value.String()
+		case outboxStateScanDelayKey:
+			stateScanDelay = v.Value.String()
+		}
+	}
+
+	if publishPubSub != "" && publishTopicKey != "" {
+		o.lock.Lock()
+		defer o.lock.Unlock()
+
+		if outboxPubsub == "" {
+			outboxPubsub = publishPubSub
+		}
+
+		if stateScanDelay == "" {
+			stateScanDelay = defaultStateScanDelay
+		}
+
+		o.outboxStores[stateStore.Name] = outboxConfig{
+			publishPubSub:  publishPubSub,
+			publishTopic:   publishTopicKey,
+			outboxPubsub:   outboxPubsub,
+			stateScanDelay: stateScanDelay,
+		}
+	}
+}
+
+// Enabled returns a bool to indicate if a state store has outbox configured
+func (o *outboxImpl) Enabled(stateStore string) bool {
+	o.lock.RLock()
+	defer o.lock.RUnlock()
+
+	_, ok := o.outboxStores[stateStore]
+	return ok
+}
+
+func transaction() (state.TransactionalStateOperation, error) {
+	uid, err := uuid.NewRandom()
+	if err != nil {
+		return nil, err
+	}
+
+	return state.SetRequest{
+		Key:   outboxStatePrefix + "-" + uid.String(),
+		Value: "0",
+	}, nil
+}
+
+// PublishInternal publishes the state to an internal topic for outbox processing
+func (o *outboxImpl) PublishInternal(ctx context.Context, stateStore string, operations []state.TransactionalStateOperation, source string) ([]state.TransactionalStateOperation, error) {
+	o.lock.RLock()
+	c, ok := o.outboxStores[stateStore]
+	o.lock.RUnlock()
+
+	if !ok {
+		return nil, fmt.Errorf("error publishing internal outbox message: could not find outbox configuration on state store %s", stateStore)
+	}
+
+	trs := make([]state.TransactionalStateOperation, 0, len(operations))
+	for _, op := range operations {
+		sr, ok := op.(state.SetRequest)
+		if ok {
+			tr, err := transaction()
+			if err != nil {
+				return nil, err
+			}
+
+			var ceData []byte
+			bt, ok := sr.Value.([]byte)
```

Comment:
the original key of the user state request is not included in the outbox event, users may want it

Reply:
That might be useful in some contexts and has come up by users already